src/sesclient.native/sesclient.native.cpp

```diff
@@ -3,89 +3,167 @@
+};
 
 struct Initializer
 {
-    Initializer()
-    {
-        int err = Microsoft::Azure::Batch::SoftwareEntitlement::Init();
-        if (err != 0)
-        {
-            throw std::runtime_error(
-                "Microsoft::Azure::Batch::SoftwareEntitlement::Init failed with error " +
-                std::to_string(err)
-            );
-        }
-    }
-
-    ~Initializer()
-    {
-        Microsoft::Azure::Batch::SoftwareEntitlement::Cleanup();
-    }
+	Initializer()
+	{
+		int err = Microsoft::Azure::Batch::SoftwareEntitlement::Init();
+		if (err != 0)
+		{
+			throw std::runtime_error(
+				"Microsoft::Azure::Batch::SoftwareEntitlement::Init failed with error " +
+				std::to_string(err)
+			);
+		}
+	}
+
+	~Initializer()
+	{
+		Microsoft::Azure::Batch::SoftwareEntitlement::Cleanup();
+	}
 };
 
+auto shouldShowUsage = false;
+auto configurationError = false;
+
+void CheckForMandatoryParameters(const std::unordered_map<std::string, std::string>& parameters)
+{
+	for (const auto& param : mandatoryParameterNames)
+	{
+		if (parameters.find(param) == parameters.end())
+		{
+			std::cout << "Missing mandatory parameter " << param << std::endl;
+			configurationError = true;
+		}
+	}
+}
+
+void CheckForExtraParameters(const std::unordered_map<std::string, std::string>& parameters)
+{
+	for (const auto& parameter : parameters)
+	{
+		if (std::find(mandatoryParameterNames.begin(), mandatoryParameterNames.end(), parameter.first) != mandatoryParameterNames.end()) {
+			continue;
+		}
+
+		if (std::find(optionalParameterNames.begin(), optionalParameterNames.end(), parameter.first) != optionalParameterNames.end()) {
+			continue;
+		}
+
+		std::cout << "Unexpected additional parameter " << parameter.second << std::endl;
+		configurationError = true;
+	}
+}
+
+std::unordered_map<std::string, std::string> ReadParameters(int argc, char** argv)
+{
+	std::unordered_map<std::string, std::string> parameters;
+
+	if ((argc % 2) == 1)
+	{
+		// We have pairs of parameters, collect them into our map
+		while (argc != 1)
+		{
+			auto value = argv[--argc];
+			auto key = argv[--argc];
+			parameters.emplace(key, value);
+		}
+	}
+
+	if (parameters.size() == 0)
+	{
+		shouldShowUsage = true;
+	}
+	else
+	{
+		CheckForMandatoryParameters(parameters);
+		CheckForExtraParameters(parameters);
+	}
+
+	return parameters;
+}
+
+void ConfigureConnection(const std::unordered_map<std::string, std::string>& parameters)
+{
+	auto thumbprintParameter = parameters.find("--thumbprint");
+	auto commonNameParameter = parameters.find("--common-name");
+
+	if (thumbprintParameter != parameters.end()
+		&& commonNameParameter != parameters.end())
+	{
+		Microsoft::Azure::Batch::SoftwareEntitlement::AddSslCertificate(
+			thumbprintParameter->second,
+			commonNameParameter->second
+		);
+	}
+	else if (thumbprintParameter == parameters.end()
+		&& commonNameParameter != parameters.end())
+	{
+		std::cout << "--thumbprint must be used when --common-name is used" << std::endl;
```

Comment:
At this point we're validating the parameters used, so it wouldn't be unusual for an issue to come up - thus I don't think an exception is appropriate.